src/BizHawk.Client.Common/ArgParser.cs

```diff
@@ -1,221 +1,235 @@
 ﻿#nullable enable
 
 using System.Collections.Generic;
+using System.CommandLine;
+using System.CommandLine.Builder;
+using System.CommandLine.Parsing;
 using System.Linq;
-using System.IO;
 using System.Net.Sockets;
 
 using BizHawk.Common.CollectionExtensions;
-using BizHawk.Common.StringExtensions;
 
 namespace BizHawk.Client.Common
 {
-	/// <summary>
-	/// Parses command line flags from a string array into various instance fields.
-	/// </summary>
-	/// <remarks>
-	/// If a flag is given multiple times, the last is taken.<br/>
-	/// If a flag that isn't recognised is given, it is parsed as a filename. As noted above, the last filename is taken.
-	/// </remarks>
+	/// <summary>Parses command-line flags into a <see cref="ParsedCLIFlags"/> struct.</summary>
 	public static class ArgParser
 	{
-		/// <exception cref="ArgParserException"><c>--socket_ip</c> passed without specifying <c>--socket_port</c> or vice-versa</exception>
-		public static void ParseArguments(out ParsedCLIFlags parsed, string[] args)
+		private sealed class BespokeOption<T> : Option<T>
```

Comment:
Can you explain why this class is necessary?

Reply:
https://github.com/TASEmulators/BizHawk/blob/dfe4a72960a5c92555667671f9cd70d1130dcf74/src/BizHawk.Client.Common/ArgParser.cs#L127-L131
(essentially `root.Options.Where(o => o is not BespokeOption<>)`; this is necessary for `--help` as you already noted in another comment)
The other way, possibly the intended way, is to positively identify the `HelpOption`, but that type isn't exposed so you'd be comparing qualified names.

Reply:
In my testing checking for `Option<>` seems to filter out the `HelpOption`, did you try that?

Reply:
Sorry, I missed this comment. What exactly did you try?
```diff
-var t = o.GetType();
-return !t.IsGenericType || t.GetGenericTypeDefinition() != typeof(BespokeOption<>);
+return o is Option<>;
```
...doesn't compile.
```csharp
return typeof(Option<>).IsInstanceOfType(o);
```
...compiles but matches nothing.

Reply:
```diff
-                               return !t.IsGenericType || t.GetGenericTypeDefinition() != typeof(BespokeOption<>); // no there is no simpler way to do this
+                               return !t.IsGenericType || t.GetGenericTypeDefinition() != typeof(Option<>); // no there is no simpler way to do this
```

Reply:
Oh... `HelpBuilder` just happens to inherit `Option` and not `Option<T>`. The latter seems to be only a thin subclass of `Option` purely for type safety, so I don't think it's a good idea to rely on `HelpBuilder`'s inheritance tree in this case.

Reply:
Even if it inherited `Option<T>` this code would work, because it compares exact types. Of course if you don't want to rely on `HelpOption` *not* being `Option<T>` instead then yeah, making your own wrapper type just for type comparison is the only way.